### PR TITLE
Add some integration tests using puppeteer

### PR DESCRIPTION
**Pull request by calixteman**
* run with `gulp integrationtest`

### CODE REVIEW

File: test/test.js [hunk 1]
```diff
@@ -768,7 +803,20 @@ function unitTestPostHandler(req, res) {
...
+async function runIntegrationTest(browserName, baseUrl) {
+  const browser = await startBrowser(browserName);
+  const { runMe } = require("./unit/integration.js");
+  try {
+    await runMe(baseUrl, browser);
+  } catch (err) {
+    console.log(err);
+  } finally {
+    browser.close();
+    process.exit(0);
```
Comment (timvandermeij) [inline test/test.js:815]: I think the exit code should be nonzero in the case of an error.

File: test/test.js [hunk 2]
```diff
@@ -790,14 +838,29 @@ async function startBrowser(browserName, startUrl) {
...
+  const options = {
     product: browserName,
     headless: false,
     defaultViewport: null,
-  });
-  const pages = await browser.pages();
-  const page = pages[0];
-  await page.goto(startUrl, { timeout: 0 });
+  };
+  if (browserName === "chrome") {
```
Comment (timvandermeij) [inline test/test.js:846]: Could you add a comment why these additional Chrome and Firefox preferences are necessary?

File: test/unit/integration.js [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+ */
+
+const assert = require("assert");
+
+const itCallbacks = [];
+function it(desc, callback) {
+  itCallbacks.push([desc, callback]);
+}
+
+const testsCallback = [];
```
Comment (timvandermeij) [inline test/unit/integration.js:23]: Maybe call this `describeCallbacks` similar to the one above?

File: test/unit/integration.js [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+function describe(desc, filename, id, callback) {
+  testsCallback.push([desc, filename, id, callback]);
+}
+
+function selector(id) {
+  return `#\\3${id.charAt(0)} ${id.slice(1)}`;
+}
+
+async function runMe(baseUrl, browser) {
```
Comment (timvandermeij) [inline test/unit/integration.js:32]: Maybe call this entrypoint `runTests` instead so it's a bit more descriptive?

File: test/unit/integration.js [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+        await page.waitForSelector(selector(id), {
+          timeout: 0,
+        });
+        itCallbacks.push(["", async () => page.bringToFront()]);
+        cb(page);
+      })()
+    )
+  );
+
+  // Execute it sequencially
```
Comment (timvandermeij) [inline test/unit/integration.js:48]: `sequencially` -> `sequentially`

File: test/unit/integration.js [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+}
+
+function selector(id) {
+  return `#\\3${id.charAt(0)} ${id.slice(1)}`;
+}
+
+async function runMe(baseUrl, browser) {
+  // Create the pages
+  await Promise.all(
+    testsCallback.map(([_, filename, id, cb]) =>
```
Comment (timvandermeij) [inline test/unit/integration.js:35]: We don't seem to be using the `id` here at all. Removing that would also make this helper more generally useful for other integration tests.

**Review by timvandermeij — COMMENTED** [review]
Nice idea! I primarily have some comments that should make this more generally usable. We should think about if we want to make this part of the bot/Travis tests already or to do that later.

File: test/test.js (hunk 2 above)
Comment (calixteman) [reply test/test.js:828]: About option for Chrome, they're required at least on my machine because without them chrome is crashing!
And as a side effect I just discovered that unit tests are running on chrome too.

File: test/unit/integration.js (hunk 6 above)
Comment (calixteman) [reply test/unit/integration.js]: the id is used as a selector to select the element to wait for before running the test.

File: test/integration/integration-boot.js [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+
+function loadTestFiles() {
+  ["./scripting.js", "./annotation.js"].map(function (moduleName) {
+    return require(moduleName);
+  });
+}
+
+const describes = [];
+
+global.describe = (desc, filename, selector, callback) => {
```
Comment (brendandahl) [inline test/integration/integration-boot.js:26]: Instead of adding a new test framework, can we load jasmine (which we already use)? I was hoping to use the integration tests for more than just form tests which this is pretty tailored too. This current method would then be a helper that a test would use.

**Review by brendandahl — CHANGES_REQUESTED** [review]
For getting various events when a page is rendered and forms are ready we can enable the dom dispatch. It's currently only enabled for development and mozcentral builds, so we'd need to add a new build flag or option to https://github.com/mozilla/pdf.js/blob/a618b02e6274c173c414855d8311930c43b99c00/web/ui_utils.js#L806

**Comment by Snuffleupagus** [comment]
> For getting various events when a page is rendered and forms are ready we can enable the dom dispatch.

While we *used* to dispatch events to the DOM, that was *purposely* removed since it's not a great API in the general case (and only existed for backwards compatibility); hence can we please *not* re-add that sort of functionality here!

If it's turns out to be *absolutely* necessary, with no way around it, please at least limit this to *only* being enabled (using pre-processor statements) for `TESTING`-builds.

> It's currently only enabled for development and mozcentral builds, so we'd need to add a new build flag or option to

That's not entirely accurate, since it's *only* enabled for mozilla-central (and nowhere else) to not break the existing tests in https://searchfox.org/mozilla-central/source/toolkit/components/pdfjs/test

**Comment by timvandermeij** [comment]
Related: https://github.com/mozilla/botio-files-pdfjs/pull/28

**Comment by brendandahl** [comment]
For the windows failure we need to increase the timeout for the beforeAll (second argument). There's suppose to be a way to change it for everything with jasmine.DEFAULT_TIMEOUT_INTERVAL but I can't seem to get that to apply.

File: gulpfile.js [hunk 8]
```diff
@@ -1502,6 +1519,31 @@ gulp.task(
...
+
+      rimraf.sync(sandboxDir);
+
+      return createSandboxBundle(defines, {
+        disableVersionInfo: true,
+      }).pipe(gulp.dest(sandboxDir));
+    }
+  )
+);
+
```
Comment (Snuffleupagus) [inline gulpfile.js:1546]: Is there a reason for moving all of this code?

File: gulpfile.js [hunk 9]
```diff
@@ -1545,6 +1589,24 @@ gulp.task(
...
+  done();
+});
+
+gulp.task(
+  "integrationtest",
+  gulp.series(
+    "integration-pre",
+    "testing-pre",
+    "generic",
+    "components",
```
Comment (Snuffleupagus) [inline gulpfile.js:1603]: Are you actually using the output of the "components" build for anything here?

File: test/integration/annotation_spec.js [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+            `${global.integrationBaseUrl}?file=/test/pdfs/annotation-highlight.pdf`
+          );
+          await page.bringToFront();
+          await page.waitForSelector("[data-annotation-id='19R']", {
+            timeout: 0,
+          });
+          return page;
+        })
+      );
+    }, 1e5);
```
Comment (Snuffleupagus) [inline test/integration/annotation_spec.js:34]: Having a *seemingly* arbitrary constant here, and in other files, is neither very clear nor maintainable in the future. Please model the new `test/integration-boot.js` file after the "boot"-files used with the existing unit- and font-tests instead, note in particular https://github.com/mozilla/pdf.js/blob/93b3ba2abb4e2eb760589899ff38cb99e8f0cbae/test/unit/jasmine-boot.js#L195-L196 and
https://github.com/mozilla/pdf.js/blob/93b3ba2abb4e2eb760589899ff38cb99e8f0cbae/test/font/jasmine-boot.js#L136-L137

**Review by Snuffleupagus — CHANGES_REQUESTED** [review]

File: web/app_options.js [hunk 11]
```diff
@@ -260,7 +260,8 @@ if (
   defaultOptions.sandboxBundleSrc = {
     /** @type {string} */
     value:
-      typeof PDFJSDev === "undefined" || !PDFJSDev.test("PRODUCTION")
+      typeof PDFJSDev === "undefined" ||
+      PDFJSDev.test("!PRODUCTION && !INTEGRATIONTEST")
```
Comment (Snuffleupagus) [inline web/app_options.js:264]: Also, are these changes actually necessary since there *should* be no build-target setting `PRODUCTION = false` anyway?

File: gulpfile.js (hunk 8 above)
Comment (calixteman) [reply gulpfile.js:1546]: Correct, it's useless now

File: gulpfile.js (hunk 9 above)
Comment (calixteman) [reply gulpfile.js:1603]: No

File: web/app_options.js (hunk 11 above)
Comment (calixteman) [reply web/app_options.js:264]: Good question, there was a `!PDFJSDev.test("PRODUCTION")`, is this value different of  `PDFJSDev.test("!PRODUCTION")` ?
And there's a similar use here:
https://github.com/mozilla/pdf.js/blob/master/web/app_options.js#L248

File: test/integration/annotation_spec.js (hunk 10 above)
Comment (calixteman) [reply test/integration/annotation_spec.js:34]: I tried to play with `jasmine.DEFAULT_TIMEOUT_INTERVAL` and it never worked.
Finally I found how to set it correctly: `jasmine.jasmine.DEFAULT_TIMEOUT_INTERVAL`.

File: web/app_options.js (hunk 11 above)
Comment (Snuffleupagus) [reply web/app_options.js:264]: > there was a `!PDFJSDev.test("PRODUCTION")`, is this value different of `PDFJSDev.test("!PRODUCTION")` ?

Those two formats are equivalent.

**Review by brendandahl — APPROVED** [review]
Let's merge this if the full test suite runs fine, so we can unbreak tests for other PR's.

**Comment by timvandermeij** [comment]
Thank you for setting this up! This will be very useful so we can actually start writing integration tests for the viewer now.
